rfcs/0003-timestamps-and-ttl.md

```diff
@@ -0,0 +1,281 @@
+flexible, however, as many rows may not be updated in a long time (though most of those rows will be
+in the base layer of the LSM tree anyway, so it may not be so bad).
+
+### SlateDB::Get
+
+SlateDB currently returns a `Option<Bytes>` on a successful read operation. Now that additional
+metadata (timestamps in this RFC) are introduced, we will need to expand this API to include
+more information in the return value. We have a few options here (TODO: update this section when
+we converge on the approach we want):
+
+1. We can return a strongly typed result (such as `Option<Row>` that is a struct with the result as
+   well as any additional retrieved metadata)
+2. We can introduce a `db#get_with_metadata` that will return a tuple of `Bytes` and corresponding
+   metadata
+3. Do nothing (make the timestamp opaque to the user and require them to interact with timestamps
+   only via the public API)
+
+## Implementation
+
+### Data Format
+
+Currently, SlateDB does not have a mechanism for backwards-compatibility of the SST data format.
+We will first introduce the following two fields to the flatbuffer definitions: 
+- `format_version`: field that specifies the codec version that encoded this SST
+- `meta_features`: an enumeration of metadata fields that are available with each encoded row. This
+  makes it possible in the future to save space per row by disabling metadata features, but that is
+  out of scope for this RFC
+
+```fbs
+enum MetaFeatures: byte {
+    Timestamp,
+    TimeToLive,
+}
+
+table SsTableInfo {
+    // ...
+    
+    // the current encoding version of the data block
+    format_version: uint;
+    
+    // the metadata features that are encoded with each row, in order that they are encoded
+    meta_features: [MetaFeatures];
+}
+```
+
+This value will be written when the SST is created. We will read this back using a codec that maps
+1:1 with the format version. The version 0 has the following per-value format:
+
+```
+|  u16    | var |    u32    |  var  |
+|---------|-----|-----------|-------|
+| key_len | key | value_len | value |
+|---------|-----|-----------|-------|
+```
+
+The `format_version` will be bumped to 1 and will be modified to have the following schema:
+
+```
+|  u16    | var | var  |    u32    |  var  |
+|---------|-----|------|-----------|-------|
+| key_len | key | meta | value_len | value |
+|---------|-----|------|-----------|-------|
+```
+
+The newly introduce `meta` field will be decoded using the `meta_features` array specified for
+this SST. In cases where both `Timestamp` and `TimeToLive` are enabled, the row will look like:
+
+```
+|  u16    | var | i64 | i64 |    u32    |  var  |
+|---------|-----|-----|-----|-----------|-------|
+| key_len | key | ts  | ttl | value_len | value |
+|---------|-----|-----|-----|-----------|-------|
+```
+
+Both `ts` and `ttl` are encoded as `i64` in milliseconds. The expiry time is `ts + ttl`.
```

Comment:
not sure whether it's better to encode it like this or as `ts, expire_ts`. It contains the same information. `expire_ts` is nice because we don't have to do the extra decode/addition per row retrieved so maybe that's better? (we could skip de-serializing the `ts` in the compaction filter, for example)